packages/server/src/model-view.ts

```diff
@@ -0,0 +1,60 @@
+import { ValueModel, ViewModel } from './types'
+
+export function view<T>(
+  loadInput: (get: <V>(model: ValueModel<V>) => V | undefined) => T
+): ViewModel<T> {
+  let state: T | undefined = undefined
+  let isInvalid = true
+  let deps = new Set<ValueModel<any>>()
+  let depSubs = new Map<ValueModel<any>, () => void>()
+  const subHandlers = new Set<(newState: T) => void>()
+  function get(): T {
+    if (!isInvalid) return state as T
+    depSubs.forEach((unsub) => unsub()) // unsubscribe from all deps
+    deps = new Set<ValueModel<any>>()
+    depSubs = new Map<ValueModel<any>, () => void>()
+    function getter<V>(model: ValueModel<V>): V | undefined {
+      if (typeof model === 'function') return model()
+      deps.add(model)
+      depSubs.set(
+        model,
+        model.subscribe(() => {
+          isInvalid = true
+        })
+      )
+      if (model.type === 'query') {
+        return model.get()
+      } else if (model.type === 'state') {
+        return model.get()
+      } else if (model.type === 'view') {
+        return model.get()
+      }
```

Comment:
Maybe we could use getModelState helper here too? Sounds like it handles this case + it's .get() for each model anyway right now.

Reply:
yeah sure